Add more overflow tests for timedelta64 operations

patrickmckenna opened the pull request:
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

---

This PR adds more specific tests, to verify the current, sometimes inconsistent behavior. Much of it has been reported/discussed in previous threads, and some appears to reflect intentional tradeoffs, e.g. allowing rounding error to avoid overflows. Other behavior seems likely unintentional:

- Inconsistent usage of `ValueError` and `OverflowError`
- Spurious `OverflowError`s for 1-element `Series` w/ valid values

Since this is my first time poking around pandas/numpy internals, it seemed best to make sure the right tests are in the right places before attempting any fixes.



### Open Questions
- What exception should be raised if `Series.sum()` would yield an invalid `Timedelta`: `ValueError`, `OutOfBoundsTimedelta`, something else? (I can rewrite the tests to mark those cases that currently raise the "wrong" exception as `xfail` for now.)
- Is silently introducing rounding error considered ok, or should a warning be emitted?
- Is this the right spot for these tests? AFAICT, a lot of this behavior is determined by `pandas.core.nanops`, particularly special-casing of `timedelta64` in `nansum()`: https://github.com/pandas-dev/pandas/blob/4cf8d55b1549be756857d82480afec7072028a73/pandas/core/nanops.py#L618-L621 My assumption was that tests against the public API would be preferred, but please LMK if that's not the case.



### Prior Art
- https://github.com/pandas-dev/pandas/issues/43178
- https://github.com/pandas-dev/pandas/issues/9442
- https://github.com/pandas-dev/pandas/pull/10926

patrickmckenna committed:
add tests for existing behavior

Currently all timedelta64 sums involve int -> float conversion.

patrickmckenna committed:
note spurious overflow for single elem td64 series

patrickmckenna committed:
finer-grained testing for td64 sum overflow errors

pep8speaks commented:
Hello @patrickmckenna! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/tests/series/test_reductions.py`](https://github.com/pandas-dev/pandas/blob/992ca95381a4dfefcc00a2a5fe108406d2438fd8/pandas/tests/series/test_reductions.py):

> [Line 90:89](https://github.com/pandas-dev/pandas/blob/992ca95381a4dfefcc00a2a5fe108406d2438fd8/pandas/tests/series/test_reductions.py#L90): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 88 characters)

patrickmckenna committed:
styling, ex msg fixes

patrickmckenna committed:
Merge remote-tracking branch 'upstream/main' into td64-sums

patrickmckenna committed:
Merge remote-tracking branch 'upstream/main' into td64-sums

patrickmckenna committed:
consolidate, parameterize td64 addition overflow tests

patrickmckenna committed:
add scalar multiplication tests

patrickmckenna committed:
add tests for scalar multiplication

patrickmckenna committed:
Merge remote-tracking branch 'upstream/main' into td64-sums

patrickmckenna committed:
mypy, win38 fixes

patrickmckenna committed:
use box_expected where possible

patrickmckenna committed:
Merge remote-tracking branch 'upstream/main' into td64-sums

patrickmckenna commented on pandas/tests/arithmetic/test_timedelta64.py:
```diff
@@ -275,6 +368,123 @@ def test_comparisons_coverage(self):
+    left = wrap_value(Timedelta.max, add_sub_types.left)
+    right = wrap_value(positive_td, add_sub_types.right)
+
+    if add_sub_types.result is Timedelta:
```
Should this be raising a `OutOfBoundsTimedelta` in all cases? Ref: https://github.com/pandas-dev/pandas/pull/34448

jbrockmendel replied:
probably, yes

patrickmckenna reviewed (commented):
This PR adds some additional overflow tests for timedeltas. It's goal is to accurately capture the current behavior, not fix existing bugs/inconsistencies&mdash;but that's the next step 😄 

@jbrockmendel as one who's been working on related issues somewhat recently, I'd appreciate your feedback! (Apologies if there are others I should be pinging, too...) This is my first time poking around pandas/numpy internals, so if there's a better way to write/organize the tests, please LMK.

jreback commented on pandas/tests/arithmetic/test_timedelta64.py:
```diff
@@ -31,12 +44,92 @@
 
+timedelta_types = (Timedelta, TimedeltaArray, TimedeltaIndex, Series, DataFrame)
+timestamp_types = (Timestamp, DatetimeArray, DatetimeIndex, Series, DataFrame)
+containers = slice(1, None)
```
ok can you move all of these changes (and the existing overflow tests) to a new file, test_timedelta64_overflow.py or similar. We use hypothesis in very specific ways and most of this is just test setup which is  confusing.

patrickmckenna replied:
> ok can you move all of these changes (and the existing overflow tests) to a new file

@jreback sure thing 👍. I've moved ([bd62e08](https://github.com/pandas-dev/pandas/pull/46854/commits/bd62e0844af1604089c7bd7e7a4bb06f0ddaf5af)) all of the new overflow tests, plus those that already existed in the 2 files this PR touches, to `pandas/tests/test_timedelta64_overflow.py`.

Some td64 overflow tests remain in other modules. Should I also move these, or should they stay in place?   
  - `tests/tslibs/test_conversion.py::test_ensure_timedelta64ns_overflows()`
  - `tests/tslibs/test_timedeltas.py::test_huge_nanoseconds_overflow()`
  - `tests/scalar/timedelta/test_timedelta.py::test_mul_preserves_reso()`
  - `tests/scalar/timedelta/test_constructors.py::test_construct_from_td64_with_unit(),test_overflow_on_construction()`

Happy to do either, just wondering what the preferred organization is? I left the above stuff in place based on the recommendation 1 in the [docs](https://github.com/pandas-dev/pandas/blob/8980af7ce9d98713b0f8792e38f0fe43088e8780/doc/source/development/test_writing.rst).


> We use hypothesis in very specific ways and most of this is just test setup which is confusing.

Yeah, the new tests probably use it gratuitously 😬 I'll pare that back and try to better match the existing usage patterns.

jbrockmendel replied:
> Some td64 overflow tests remain in other modules. Should I also move these, or should they stay in place?

these should stay in place

jreback reviewed (changes requested):
thanks for the tests. these look to be nicely comprehensive and you have a lot of questions about these as we may have tradeoffs in the existing operations. If you can move these to a new file will make it easier to run.

patrickmckenna committed:
consolidate all new, some old td64 overflow tests in new module

Some td64 overflow tests remain in other modules:

  - tests/tslibs/test_conversion.py::test_ensure_timedelta64ns_overflows()
  - tests/tslibs/test_timedeltas.py::test_huge_nanoseconds_overflow()
  - tests/scalar/timedelta/test_timedelta.py::test_mul_preserves_reso()
  - tests/scalar/timedelta/test_constructors.py::test_construct_from_td64_with_unit(),test_overflow_on_construction()

Still TBD whether these should remain there or also be migrated.

See: github.com/pandas-dev/pandas/pull/46854#discussion_r858131625

jbrockmendel commented:
i'll take a look today

patrickmckenna committed:
remove hypothesis overuse

jbrockmendel commented on pandas/tests/series/test_reductions.py:
```diff
@@ -51,29 +53,32 @@ def test_td64_sum_empty(skipna):
+@given(
+    st.integers(
+        min_value=0,
+        max_value=10 ** (np.finfo(np.float64).precision),
```
any reason not to write `10**15` explicitly?

patrickmckenna replied:
Nope, will do.

jbrockmendel commented on pandas/tests/test_timedelta64_overflow.py:
```diff
@@ -0,0 +1,322 @@
+#     test_overflow_on_construction()
+
+
+class BinaryOpTypes(NamedTuple):
```
this isn't a pattern we use anywhere else.  is there a viable way to make this more idiomatic?

patrickmckenna replied:
Yeah, think so. I'll refactor the tests.

jbrockmendel commented on pandas/tests/test_timedelta64_overflow.py:
```diff
@@ -0,0 +1,322 @@
+    TimedeltaArray,
+)
+
+timedelta_types = (Timedelta, TimedeltaArray, TimedeltaIndex, Series, DataFrame)
```
aside from the scalar Timedelta, this could use the box_with_array fixture in tests.arithmetic (we could plausibly move that fixture to a higher level)

patrickmckenna replied:
Thanks for the tip, switched to `box_with_array` in most cases. (That fixture's in the top-level `conftest`, so think we're good for now.)

jbrockmendel commented on pandas/tests/test_timedelta64_overflow.py:
```diff
@@ -0,0 +1,322 @@
+    ).map(Timedelta)
+)
+def test_td64_summation_raises_value_error_for_most_overflows(value: Timedelta):
+    s = Series([Timedelta.max, value])
```
nitpick, pls avoid 1-letter variable names.  e.g. `ser` instead of `s`

jbrockmendel commented on pandas/tests/test_timedelta64_overflow.py:
```diff
@@ -0,0 +1,322 @@
+get_item_names = lambda t: "-".join(map(attrgetter("__name__"), t))
+
+
+# TODO: consolidate remaining td64 overflow tests here?
```
no.

jbrockmendel commented on pandas/tests/test_timedelta64_overflow.py:
```diff
@@ -0,0 +1,322 @@
+        value * td
+
+
+class TestAddSubNaTMasking:
```
this class is just moved but unchanged?

patrickmckenna replied:
Yeah, that was the case, but based on https://github.com/pandas-dev/pandas/pull/46854#issuecomment-1110507748 I returned it to its original home.

jbrockmendel commented on pandas/tests/test_timedelta64_overflow.py:
```diff
@@ -0,0 +1,322 @@
+@given(value=st.floats().filter(lambda f: abs(f) > 1))
+@pytest.mark.parametrize(
+    argnames="td_type",
+    argvalues=timedelta_types[containers],  # type: ignore[arg-type]
```
this is a pattern im not familar with -> mental overhead.  is this useful enough to ask reviewers/future-contributors to learn it?  or can we use a more plain-vanilla pattern?

patrickmckenna replied:
Oh no, I'm not trying to advocate for a new pattern (think this just matched my mental model at the time). I'll rewrite this to better follow existing conventions.

jbrockmendel commented on pandas/tests/test_timedelta64_overflow.py:
```diff
@@ -0,0 +1,322 @@
+    left = wrap_value(Timestamp.min, ts_add_sub_types[1])
+
+    ex = (OutOfBoundsDatetime, OverflowError)
+    msg = "|".join(["Out of bounds nanosecond timestamp", "Overflow in int64 addition"])
```
the "int64 addition" message should probably be caught and changed to something more appropriate

patrickmckenna replied:
Agreed. Imagine this would be part of the same future PR mentioned in https://github.com/pandas-dev/pandas/pull/46854#discussion_r861277416?

jbrockmendel commented on pandas/tests/test_timedelta64_overflow.py:
```diff
@@ -0,0 +1,322 @@
+    else:
+        box_cls = type_
+
+    return type_(tm.box_expected([value], box_cls))
```
is the `type_` here necessary?  can you call it `cls` instead?

jbrockmendel commented on pandas/tests/test_timedelta64_overflow.py:
```diff
@@ -0,0 +1,322 @@
+):
+    s = Series(value)
+
+    msg = "int too big to convert|Python int too large to convert to C long"
```
nitpick: can you write these as `msg = "|".join(["foo", "bar"])` instead of `msg = "foo|bar"`

jbrockmendel commented on pandas/tests/test_timedelta64_overflow.py:
```diff
@@ -0,0 +1,322 @@
+        [
+            "cannot convert float infinity to integer",
+            "Python int too large to convert to C long",
+            "int too big to convert",
```
in most of these cases we probably want these to raise OutOfBoundsTImedelta with a timedelta-specific exception message

patrickmckenna replied:
That seems reasonable. AFAICT, the key to getting there is to catch `OverflowError`s here: https://github.com/pandas-dev/pandas/blob/9f2491875945d4b515f66491dbe507632060d6bf/pandas/_libs/tslibs/timedeltas.pyx#L264

Once that's one, I *think* we'd remove this check, too: https://github.com/pandas-dev/pandas/blob/9f2491875945d4b515f66491dbe507632060d6bf/pandas/core/nanops.py#L378-L380

Does that seem right, or am I missing something? I can open a follow-up PR for that work.

jbrockmendel commented on pandas/tests/test_timedelta64_overflow.py:
```diff
@@ -0,0 +1,322 @@
+        left - right
+
+
+@given(value=st.floats().filter(lambda f: abs(f) > 1))
```
is this really an improvement over e.g. `value = np.random.random()**2`?  I don't have a strong opinion, mostly curious.  how much test-runtime overhead does using hypothesis for this add?

patrickmckenna replied:
Heh, unsure, but also wonder the same. For simplicity, I switched to a few hardcoded values for now, but am curious to look into this more in the future (or hear from someone who has).

jbrockmendel commented on pandas/tests/series/test_reductions.py:
```diff
@@ -51,29 +53,32 @@ def test_td64_sum_empty(skipna):
-        (s1 - s1.min()).sum()
-    s2 = ser[0:1000]
-    (s2 - s2.min()).sum()
+    assert result != value
```
we probably still want to assert that the result is _close_ to correct right?

jbrockmendel commented on pandas/tests/series/test_reductions.py:
```diff
@@ -51,29 +53,32 @@ def test_td64_sum_empty(skipna):
+    assert result == value
 
-    # the computation is converted to float so
-    # might be some loss of precision
```
i think itd be useful to retain this comment

jbrockmendel replied:
i see you moved this comment to a docstring, thanks.

For future reference, pls let the commenter hit the "resolve" button

jbrockmendel commented:
My inclination is to avoid making the new file, keep reduction tests in test_reductions, arithmetic tests in tests.arithmetic, scalar tests in tests.scalar.timedelta

jbrockmendel commented on pandas/tests/arithmetic/test_timedelta64.py:
```diff
@@ -31,12 +44,87 @@
     get_upcast_box,
 )
 
+timedelta_types = (Timedelta, TimedeltaArray, TimedeltaIndex, Series, DataFrame)
```
that assumption is accurate

id prefer to keep the Timedelta scalar tests in tests.scalar.timedelta, then use the box_with_array fixture for the non-scalar tests.  part of the virtue of this is that when writing the non-scalar tests, we assume that the scalar behavior is correct and tested.

jbrockmendel commented:
OK, i've done a lot of nitpicking, time for some positive reinforcement.  thanks for the PR!  getting consistent exceptions and messages is not easy, glad to have you working on it

patrickmckenna committed:
Merge remote-tracking branch 'upstream/main' into td64-sums

patrickmckenna committed:
address PR code style feedback

patrickmckenna committed:
put new tests back in pre-existing modules

Still need to DRY up a few things.

patrickmckenna committed:
dedupe test setup helpers

patrickmckenna committed:
Merge remote-tracking branch 'upstream/main' into td64-sums

patrickmckenna committed:
obey the linting gods

patrickmckenna commented:
> OK, i've done a lot of nitpicking, time for some positive reinforcement. thanks for the PR! getting consistent exceptions and messages is not easy, glad to have you working on it

Happy to help, and thanks for the specific feedback! (FWIW, I appreciate the detailed guidance, especially when jumping into a new-to-me codebase 😀)

patrickmckenna committed:
Merge remote-tracking branch 'upstream/main' into td64-sums

patrickmckenna committed:
adjust for platform/env-specific overflow behavior

patrickmckenna commented on pandas/_testing/__init__.py:
```diff
@@ -279,6 +281,28 @@ def box_expected(expected, box_cls, transpose=True):
     return expected
 
 
+def wrap_value(value, cls, transpose=False):
```
The motivation for this was to simplify writing tests that should produce the same output for both scalar and container types, e.g. [here](https://github.com/pandas-dev/pandas/pull/46854/files#diff-fbcbe1c510363cf465fa8eb7f5cf8fc3f6bfcc9e1fab7b93f1b69521de44d194R2209).

jbrockmendel replied:
are there any other thematically similar functions that could be grouped together in a pandas._testing.foo instead of `__init__`?  we've kind of stagnated, but in principle there's a goal to get stuff out of this file

patrickmckenna commented on pandas/tests/arithmetic/test_timedelta64.py:
```diff
@@ -2089,18 +2148,104 @@ def test_td64arr_pow_invalid(self, scalar_td, box_with_array):
-    timestamp = Timestamp("2021-01-01")
-    result = timestamp + timedelta_range("0s", "1s", periods=31)
-    expected = DatetimeIndex(
+class TestAddSub:
```
If this organization makes sense, I can open a follow up PR to migrate many of the existing tests into this and similar classes (and make them fully parameterized across all container types).

patrickmckenna reviewed (commented):
@jbrockmendel @jreback per https://github.com/pandas-dev/pandas/pull/46854#issuecomment-1110507748, I've migrated the new tests back to existing modules. I've also removed the hypothesis usage, and rewritten the tests to follow the pattern of accepting box types as fixtures. Please LMK if you see anything else that needs changing. 

Once this ships, I can open up [a follow-on PR](https://github.com/pandas-dev/pandas/pull/46854#discussion_r861277416) to make the exception raising more consistent.

patrickmckenna commented on pandas/tests/arithmetic/test_timedelta64.py:
```diff
@@ -2089,18 +2148,104 @@ def test_td64arr_pow_invalid(self, scalar_td, box_with_array):
+        ],
     )
-    tm.assert_index_equal(result, expected)
+    def test_sub_raises_if_result_would_overflow(
```
FWIW, I tried rewriting this to keep the test function body constant, and mapping each input value to an appropriate `xfail` marker:

```python
xfail_does_not_raise = pytest.mark.xfail(reason="doesn't raise", strict=True)
xfail_raises_overflow_error = pytest.mark.xfail(
    reason="raises wrong error",
    raises=OverflowError,
    strict=True,
)
# ...
    @pytest.mark.parametrize(
        "rval",
        [
            pytest.param(Timedelta(1), marks=xfail_does_not_raise),
            pytest.param(Timedelta(2), marks=xfail_raises_overflow_error),
            pytest.param(Timedelta.max, marks=xfail_raises_overflow_error),
        ],
    )    
    def test_sub_raises_td64_specific_error_if_result_would_overflow(
        self,
        max_td64: TD64_BOX_TYPE,
        rval: Timedelta,
        td64_type: Type[TD64_TYPE],
    ):
        rvalue = tm.wrap_value(rval, td64_type)
        min_td64 = -1 * max_td64

        with pytest.raises(OutOfBoundsTimedelta, match="too small"):
            min_td64 - rvalue
```

Unfortunately that slowed execution by a factor of ~40 (locally). Decorating the test with a single, catch-all `xfail` marker led to the same results. Passing `run=False` did help, but it was still ~5x slower.

jbrockmendel commented:
will take another look tomorrow

jbrockmendel commented on pandas/tests/arithmetic/test_timedelta64.py:
```diff
@@ -1,9 +1,18 @@
+from functools import partial
+from typing import (
+    Type,
+    Union,
```
We can't use `type` and `|`?

patrickmckenna replied:
I wish, that's [a `3.10` feature](https://docs.python.org/3/whatsnew/3.10.html#pep-604-new-type-union-operator).

jbrockmendel replied:
even if we do `from __future__ import annotations`?  im sure we use pipes elsewhere

patrickmckenna replied:
Huh, I'd tried that and switching to pipes, and it was failing for type aliases like `TD64_BOX_TYPE = TimedeltaArray | TimedeltaIndex | Series | DataFrame`. But TIL: pipes will work for function annotations (with that `import`).

`type` does work. I'll switch to that, and replace the aliases with pipes.

patrickmckenna replied:
Huh, I'd tried that and switching to pipes, and it was failing for type aliases like `TD64_BOX_TYPE = TimedeltaArray | TimedeltaIndex | Series | DataFrame`. But TIL: pipes will work for function annotations (with that `import`).

`type` does work. I'll switch to that, and replace the aliases with pipes.

jbrockmendel commented on pandas/tests/arithmetic/test_timedelta64.py:
```diff
@@ -59,6 +89,59 @@ def get_expected_name(box, names):
+    """
+    dt64 = tm.wrap_value(Timestamp.now(), dt64_type)
+    td64 = tm.wrap_value(Timedelta(0), td64_type)
+    return type(dt64 + td64)
```
so this is going to map Timedelta->Timestamp and TimedeltaArray->DatetimeArray but be an identity mapping for everything else?

patrickmckenna replied:
I might be misunderstanding the question, but think the gist of what you're suggesting is right (assuming Array stands for every box type). So:

```
Timedelta + Timestamp -> Timestamp
Timedelta + DatetimeArray -> DatetimeArray
Timedelta + DatetimeIndex -> DatetimeIndex
Timedelta + Series -> Series
Timedelta + DataFrame -> DataFrame
TimedeltaArray + Timestamp -> DatetimeArray
TimedeltaArray + DatetimeArray -> DatetimeArray
...
TimedeltaIndex + Timestamp -> DatetimeIndex
TimedeltaIndex + DatetimeArray -> DatetimeIndex
...
```

Happy to update the docstring, or use a different pattern, if that's unclear.

jbrockmendel replied:
My preference (mentioned https://github.com/pandas-dev/pandas/pull/46854#issuecomment-1110507748) is to handle the Timedelta/Timestamp cases in the existing tests/scalar/... files, and the non-scalar tests here.  Then this wrapping becomes unnecessary and you can just use box_with_array

patrickmckenna replied:
How about scalar/box tests, e.g. `Timedelta + DatetimeIndex`, where should those live?

patrickmckenna committed:
Merge remote-tracking branch 'upstream/main' into td64-sums

patrickmckenna committed:
use newer type hint syntax

patrickmckenna committed:
update tests to reflect recent changes

patrickmckenna committed:
remove scalar-box tests, use existing fixtures

patrickmckenna committed:
DRY up some td64 tests

patrickmckenna committed:
platform-specific fixes

patrickmckenna commented:
> happy for you to merge this when you are ready

@jreback appreciate that! Alas, no green button for me&mdash;I haven't got write access 🙃

jreback commented:
> > happy for you to merge this when you are ready
> 
> @jreback appreciate that! Alas, no green button for me—I haven't got write access 🙃

@patrickmckenna that was to @jbrockmendel  :->

patrickmckenna commented:
Hah, whoops! 🤦‍♂️😅

patrickmckenna committed:
Merge branch 'main' into td64-sums

patrickmckenna committed:
Merge branch 'main' into td64-sums

patrickmckenna commented:
Please LMK if there's more that needs doing here. Unsure what to do about the few red x's, which appear to be (transient?) failures during CI setup.

patrickmckenna committed:
Merge remote-tracking branch 'upstream/main' into td64-sums

patrickmckenna committed:
Merge remote-tracking branch 'upstream/main' into td64-sums

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
